forem/forem
# Admin Panel for Configurable Navigational Links

## What type of PR is this? (check all applicable)


## Description
_(Backend Only, Frontend PR in a subsequent PR)_

Currently we have a couple "key links" areas. (left sidebar on home page, site footer, key links page.

We should be able to define what these links are at the app level, both for generalization purposes and also possibly for dev.to administration simplification. Admins of Forems  should have the ability to add to the navigation sidebar links with additional community products such as books, movies etc., whilst being able to remove the links that are specific to DEV or those that they do not want.

The configurable elements of a navigation item is:
- Friendly name
- Icon (the podcast icon as an example).
- The destination page url
- A value that sets whether the link should be shown in the sidebar 'pre auth' or not.

You'll see that this PR was a journey and a **huge learning experience** for me:🛣
 At first I tried to fit the navigation links within the Site Config page and model, however as I started working with the more complicated actions like "update", I began to understand @citizen428's foresighted early-on comments (see below for comments). The navigation links are not simple scalar values or key-value pairs, but it is essentially an array of hashes which adds unnecessary complexity to work with within the Site Config. 
Hence it made more sense to pull the configurable Navigation Links into it's own page (in the future we'll probably move more sections out of the site config as its getting too complicated) and backed by its own table. In this way we could also add validations on the model without needing to go through a service.

This PR covers the admin portion of adding the database table + the admin page to configure the links. A subsequent PR will remove the hardcoded links and replace them with the values form the table. I chose to separate out the PR's so we can prep the database for DEV before we update the frontend :) 

## Related Tickets & Documents
https://github.com/forem/forem/issues/4661
https://github.com/forem/forem/issues/5475

## QA Instructions, Screenshots, Recordings
1. Go to http://localhost:3000/admin/navigation_links
2. Play around with the CRUD
3. Check that it saves correctly to the db 
 
**View:**
<img width="1380" alt="Screenshot 2020-10-01 at 16 07 26" src="https://user-images.githubusercontent.com/2786819/94819994-3d4e8800-0400-11eb-86c5-cc0df0ccb051.png">

**Create:**
<img width="1133" alt="Screenshot 2020-10-01 at 16 37 55" src="https://user-images.githubusercontent.com/2786819/94823779-87396d00-0404-11eb-904f-55ab6b00a75c.png">

**Update:**
<img width="1136" alt="Screenshot 2020-10-01 at 16 37 48" src="https://user-images.githubusercontent.com/2786819/94823761-7ee13200-0404-11eb-927c-c85a32137ef8.png">

**Delete:**
There will be a modal to verify deletion.

## Added tests?


## Added to documentation?


## [optional] Are there any post deployment tasks we need to perform?
Not right now

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](https://media.giphy.com/media/UQOjUuqeyIjAzJJ169/giphy.gif)
